src/spec.rs

```diff
@@ -0,0 +1,179 @@
+//! The contents of this module require the rustgc feature. To avoid polluting
+//! lib.rs with conditional compilation attributes, it is only exported if the
+//! rustgc feature is enabled.
+use core::{
+    alloc::{AllocError, Layout},
+    gc::{Conservative, NoTrace},
+    marker::PhantomData,
+    ptr::NonNull,
+};
+
+/// An allocation helper to potentially optimise the marking of allocated
+/// blocks.
+///
+/// Allocation which is done through this bottleneck will try to use type
+/// information to allocate blocks in pools which are better optimised for
+/// marking. Allocation is a hot path, so to reduce branching at runtime, it
+/// makes heavy use of specialization, hence the name.
+///
+/// There are three ways a block can be allocated:
+///
+///    Atomically - this means that a block contains no pointers which need
+///    tracing during marking. An atomically allocated block is a leaf node in
+///    the object graph.
+///
+///    Precisely - blocks are allocated along with a bitmap which describes
+///    where the pointers are on a per-word basis. The parts of a block which
+///    are not described in the bitmap are ignored by the collector during
+///    marking.
+///
+///    Conservatively - the whereabouts of pointers inside the block are
+///    unknown, and each word in the block must be considered for potential
+///    pointers.
+///
+/// There is a direct trade-off between the safety and performance of these
+/// three allocation types: atomic allocation will have the fastest marking
+/// times, but is the most dangerous; whereas conservative allocation is safe to
+/// use anywhere, but will have the slowest marking times.
+///
+/// Specialized allocation can be thought of as "compile-time cascading if-else"
+/// which first tries to allocate atomically, then precisely, and then finally
+/// conservatively if neither of these optimised strategies are possible.
```

Comment:
Ah, this sentence on ordering is misleading. I will remove it as I think it confuses matters.

Reply:
Fixed: d3fafa159230d25b3fe85b259eb36a6da215df5f